src/Disks/ObjectStorages/ReadOnlyMetadataStorage.h

```diff
@@ -0,0 +1,165 @@
+    }
+    void createMetadataFile(const std::string & /* path */, const std::string & /* blob_name */, uint64_t /* size_in_bytes */) override
+    {
+        /// Noop
+    }
+
+    ///
+    /// Throws
+    ///
+    void writeStringToFile(const std::string & /* path */, const std::string & /* data */) override
+    {
+        throwNotAllowed();
+    }
+    void setLastModified(const std::string & /* path */, const Poco::Timestamp & /* timestamp */) override
+    {
+        throwNotAllowed();
+    }
+    void chmod(const String & /* path */, mode_t /* mode */) override
+    {
+        throwNotAllowed();
+    }
+    void setReadOnly(const std::string & /* path */) override
+    {
+        throwNotAllowed();
+    }
+    void unlinkFile(const std::string & /* path */) override
+    {
+        throwNotAllowed();
+    }
+    void removeDirectory(const std::string & /* path */) override
+    {
+        throwNotAllowed();
+    }
+    void removeRecursive(const std::string & /* path */) override
+    {
+        throwNotAllowed();
+    }
+    void createHardLink(const std::string & /* path_from */, const std::string & /* path_to */) override
+    {
+        throwNotAllowed();
+    }
+    void moveFile(const std::string & /* path_from */, const std::string & /* path_to */) override
+    {
+        throwNotAllowed();
+    }
+    void moveDirectory(const std::string & /* path_from */, const std::string & /* path_to */) override
+    {
+        throwNotAllowed();
+    }
+    void replaceFile(const std::string & /* path_from */, const std::string & /* path_to */) override
+    {
+        throwNotAllowed();
+    }
+    void createDirectory(const std::string & /* path */) override
+    {
+        throwNotAllowed();
+    }
+    void createDirectoryRecursive(const std::string & /* path */) override
+    {
+        throwNotAllowed();
+    }
+    void addBlobToMetadata(const std::string & /* path */, const std::string & /* blob_name */, uint64_t /* size_in_bytes */) override
+    {
+        throwNotAllowed();
+    }
+    void unlinkMetadata(const std::string & /* path */) override
+    {
+        throwNotAllowed();
+    }
+
+
+    ///
+    /// Others
+    ///
+    bool supportsChmod() const override { return false; }
+
+private:
+    [[noreturn]] static void throwNotAllowed()
+    {
+        throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Only read-only transaction operations are supported");
+    }
+};
+
+/// Readonly, throws NOT_IMPLEMENTED error.
+/// Can be used to add limited read-only support of MergeTree.
+class ReadOnlyMetadataStorage : public IMetadataStorage
+{
+public:
+    ///
+    /// Noop
+    ///
+    Poco::Timestamp getLastModified(const std::string & /* path */) const override
+    {
+        /// Required by MergeTree
+        return {};
+    }
+    time_t getLastChanged(const std::string & /* path */) const override
+    {
+        return {};
+    }
+
+    struct stat stat(const String & /* path */) const override
```

Comment:
I prefer to throw NOT_IMPLEMENTED for every API that is not required for now, since who knows who can rely on this...

Reply:
It's perfectly fine to throw NOT_IMPLEMENTED when you don't have an implementation. It just looks confusing because for example `getLastChanged()` returns 0 and `readFileToString()` throws an exception. Why is that difference?

Reply:
This was based on `MetadataStorageFromStaticFilesWebServer`, and for it `getLastChanged` returns `0`.
But AFAICS MergeTree code relies on it, though for these type of disks (s3_plain and web) it should not work anyway and returning 0 is not good since this breaks the behaviour.

Reply:
While for example getLastModified required by MergeTree to work, even for read-only.

Reply:
The main problem with those two classes `ReadOnlyMetadataStorage` and `ReadOnlyMetadataTransaction` is that they're not consistent with themselves or at least with what seems from their names. I'd prefer them to be like this:

```
class ReadOnlyMetadataTransaction
{
   public:
      // all state-changing functions throw exception "that work because object storage is readonly"
      // maybe except commit (because there is always nothing to commit for this kind of transaction
};

class ReadOnlyMetadataStorage
{
   public:
      // all state-changing functions throw exception "that doesn't work because object storage is readonly"

      // state-reading functions are either left pure virtual or (if it's really necessary) throw exception 
      // but with a different text (just "that is not implemented", not like the message we throw for state-changing functions)

      // and if you want some state-reading or state-changing function to be something else
      // it can be done by overriding functions in inherited class
};
```